# [M3S02] - Ex. 3 — Métricas de Avaliação do Modelo

**Objetivo:** treinar o modelo de regressão linear múltipla e calcular as métricas MAE, RMSE e R².

**Base de dados:** `dados_AP - [M3S01] - Ex. 1 até [M3S03] - Ex. 4.csv`

In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

## 1. Preparação dos dados

In [2]:
df = pd.read_csv('../../data/input/csv/dados_AP - [M3S01] - Ex. 1 até [M3S03] - Ex. 4.csv')

X = df[['investimento_marketing', 'preco_medicamento', 'mes']]
y = df['demanda']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## 2. Treinamento do modelo

In [3]:
modelo = LinearRegression()
modelo.fit(X_train, y_train)

y_pred = modelo.predict(X_test)

## 3. Métricas de avaliação

In [4]:
mae  = mean_absolute_error(y_test, y_pred)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2   = r2_score(y_test, y_pred)

print(f'MAE  (Erro Médio Absoluto)        : {mae:.2f} unidades')
print(f'RMSE (Raiz do Erro Quadrático)    : {rmse:.2f} unidades')
print(f'R²   (Coeficiente de Determinação): {r2:.4f}')

MAE  (Erro Médio Absoluto)        : 18.70 unidades
RMSE (Raiz do Erro Quadrático)    : 28.01 unidades
R²   (Coeficiente de Determinação): 0.9544


## 4. Interpretação das métricas

| Métrica | Valor | O que significa |
|---|---|---|
| **MAE** | ~18.70 unidades | O modelo erra em média 18.70 unidades de demanda |
| **RMSE** | ~28.01 unidades | Penalizando erros maiores, o desvio sobe para 28 unidades |
| **R²** | ~0.9544 | O modelo explica **95.44%** da variação na demanda |

### O modelo apresenta bom desempenho?

**Sim, excelente.** Com R² de 0.95, o modelo captura quase toda a variação da demanda a partir das três variáveis preditoras. O MAE de 18.70 sobre uma demanda média de ~194 unidades representa um erro relativo de apenas ~9.6% — muito satisfatório para planejamento de estoque.

O RMSE (28.01) é moderadamente maior que o MAE (18.70), o que indica a presença de alguns poucos erros maiores, mas sem impacto severo na qualidade geral do modelo.

### Qual métrica é mais relevante nesse caso?

O **RMSE** é a mais relevante neste contexto farmacêutico. Como erros grandes impactam diretamente o planejamento de estoque — seja por falta de produto ou por excesso —, penalizar desvios maiores faz sentido. O RMSE reflete esse risco melhor do que o MAE, que trata todos os erros igualmente.